paddle/fluid/operators/class_center_sample_op.cu

```diff
@@ -0,0 +1,488 @@
+    auto* label = ctx.Input<Tensor>("Label");
+    auto* remapped_label = ctx.Output<Tensor>("RemappedLabel");
+    auto* sampled_local_class_center =
+        ctx.Output<Tensor>("SampledLocalClassCenter");
+    int num_classes = ctx.Attr<int>("num_classes");
+    int num_sample = ctx.Attr<int>("num_sample");
+
+    int rid = ctx.Attr<int>("ring_id");
+    int nranks = ctx.Attr<int>("nranks");
+    int rank = ctx.Attr<int>("rank");
+
+    int seed = ctx.Attr<int>("seed");
+    bool fix_seed = ctx.Attr<bool>("fix_seed");
+    PADDLE_ENFORCE_GT(num_classes, 0,
+                      platform::errors::InvalidArgument(
+                          "The value 'num_classes' for Op(class_center_sample) "
+                          "must be greater than 0, "
+                          "but the value given is %d.",
+                          num_classes));
+
+    PADDLE_ENFORCE_GT(num_sample, 0,
+                      platform::errors::InvalidArgument(
+                          "The value 'num_sample' for Op(class_center_sample) "
+                          "must be greater than 0, "
+                          "but the value given is %d.",
+                          num_sample));
+
+    PADDLE_ENFORCE_LE(num_sample, num_classes,
+                      platform::errors::InvalidArgument(
+                          "The value 'num_sample' for Op(class_center_sample) "
+                          "must be less than %d, "
+                          "but the value given is %d.",
+                          num_classes, num_sample));
+
+    auto& dev_ctx = ctx.template device_context<DeviceContext>();
+    auto place = BOOST_GET_CONST(platform::CUDAPlace, dev_ctx.GetPlace());
+
+    int batch_size = label->numel();
+
+    // step 1: Calculate num classes per device using nccl all reduce
+    std::vector<T> shard_dim_vec(nranks + 1, 0);
+    shard_dim_vec[rank + 1] = num_classes;
+
+    Tensor num_classes_per_device;
+    framework::TensorFromVector(shard_dim_vec, ctx.cuda_device_context(),
+                                &num_classes_per_device);
+    T* num_classes_per_device_ptr = num_classes_per_device.data<T>();
+
+#if defined(PADDLE_WITH_NCCL) || defined(PADDLE_WITH_RCCL)
+    if (nranks > 1) {
+      const auto& comm =
+          platform::NCCLCommContext::Instance().Get(rid, ctx.GetPlace());
+      // use global calculate stream
+      const auto calcu_stream =
+          static_cast<platform::CUDADeviceContext*>(
+              platform::DeviceContextPool::Instance().Get(ctx.GetPlace()))
+              ->stream();
+      PADDLE_ENFORCE_CUDA_SUCCESS(platform::dynload::ncclAllReduce(
+          num_classes_per_device_ptr, num_classes_per_device_ptr,
+          num_classes_per_device.numel(),
+          platform::ToNCCLDataType(num_classes_per_device.type()), ncclSum,
+          comm->comm(), calcu_stream));
+    }
+#endif
+
+    // step 2: Determine temporary device storage requirements
```

Comment:
You'd better to give an illustration on how the algorithm works in the pr description.